pyfolio/round_trips.py

```diff
@@ -13,168 +13,253 @@
-
-        assert np.all(np.abs(np.diff(np.sign(trans_sym.amount.cumsum()))) != 2)
-        trans_split.append(trans_sym)
-
-    transactions_split = pd.concat(trans_split)
-
-    return transactions_split
+        price_stack = deque()
+        dt_stack = deque()
+        for dt, t in trans_sym.iterrows():
+            if t.price < 0:
+                warnings.warn('Negative price detected, ignoring for'
+                              'round-trip.')
+                continue
+            signed_price = t.price * np.sign(t.amount)
+            abs_amount = int(abs(t.amount))
+            indiv_prices = [signed_price] * abs_amount
+            if (len(price_stack) == 0) or \
+               (copysign(1, price_stack[-1]) == copysign(1, t.amount)):
+                price_stack.extend(indiv_prices)
+                dt_stack.extend([dt] * len(indiv_prices))
+            else:
+                # Close round-trip
+                pnl = 0
+                invested = 0
+                cur_open_dts = []
+
+                for price in indiv_prices:
+                    if len(price_stack) != 0 and \
+                       (copysign(1, price_stack[-1]) != copysign(1, price)):
+                        # Retrieve first dt, stock-price pair from
+                        # stack
+                        prev_price = price_stack.popleft()
+                        prev_dt = dt_stack.popleft()
+
+                        pnl += -(price + prev_price)
+                        cur_open_dts.append(prev_dt)
+                        invested += abs(prev_price)
+
+                    else:
+                        # Push additional stock-prices onto stack
+                        price_stack.append(price)
+                        dt_stack.append(dt)
+
+                roundtrips.append({'pnl': pnl,
+                                   'open_dt': cur_open_dts[0],
+                                   'close_dt': dt,
+                                   'long': price < 0,
+                                   'rt_returns': pnl / invested,
+                                   'symbol': sym,
+                                   })
+
+    roundtrips = pd.DataFrame(roundtrips)
+
+    roundtrips['duration'] = roundtrips['close_dt'] - roundtrips['open_dt']
+
+    if portfolio_value is not None:
+        # Need to normalize so that we can join
+        pv = pd.DataFrame(portfolio_value,
+                          columns=['portfolio_value'])\
+            .assign(date=portfolio_value.index)
+        roundtrips['date'] = roundtrips.close_dt.apply(lambda x:
+                                                       x.replace(hour=0,
+                                                                 minute=0,
+                                                                 second=0))
+
+        tmp = roundtrips.assign(date=roundtrips.close_dt)\
+                        .join(pv, on='date', lsuffix='_')
+
+        roundtrips['returns'] = tmp.pnl / tmp.portfolio_value
```

Comment:
Do we want `tmp.pnl / tmp.porftolio_value.shift()`?

Alternatively, you could pass in `positions.sum(axis='columns') / (1+returns)` to set portfolio_value as the opening portfolio value. This way, the first day's return isn't NaN.